### ISSUE TITLE
Liquid Haskell approves eta-expanded function and errors on pointfree version of the same function

### ISSUE THREAD

**Issue by Gabriel439** [ucsd-progsys/liquidhaskell#675]
In the following code, `extract` type-checks fine.  Liquid Haskell returns with no errors:

```haskell
import Data.ByteString
import Data.ByteString.Unsafe

{-@ measure bslen :: ByteString -> Int @-}

{-@ unsafeTake :: n : Int -> ibs : { ibs : ByteString | bslen ibs >= n } -> { obs : ByteString | bslen obs == n } @-}

{-@ unsafeDrop :: n : Int -> ibs : { ibs : ByteString | bslen ibs >= n } -> { obs : ByteString | bslen obs == bslen ibs - n } @-}

{-@ extract :: ibs : { ibs : ByteString | bslen ibs >= 100 } -> { obs : ByteString | bslen obs == 4 } @-}
extract :: ByteString -> ByteString
extract ibs = unsafeTake 4 (unsafeDrop 96 ibs)
```

However, if we modify `extract` so that it looks like:

```haskell
extract = unsafeTake 4 . unsafeDrop 96
```

... it returns the following error:

```
**** RESULT: UNSAFE ************************************************************


 test2.hs:12:11-22: Error: Liquid Type Mismatch
 
 12 | extract = unsafeTake 4 . unsafeDrop 96
                ^^^^^^^^^^^^
 
   Inferred type
     VV : Data.ByteString.Internal.ByteString
  
   not a subtype of Required type
     VV : {VV : Data.ByteString.Internal.ByteString | bslen VV >= ?a}
  
   In Context
     ?a := {?a : GHC.Types.Int | ?a == (4 : int)}
```

**Comment by ranjitjhala** [comment]
Hi @Gabriel439, 

TL;DR -- this is a consequence of how inference works so, unfortunately 
you either (1) need extra qualifiers *or* (2) can use `--scrape-imports`. 
Fortunately we are working on it in an experimental branch, so hopefully 
will not need those in (the near?) future!

Details: 

This is a consequence of the current inference algorithm, here's a simpler variant:

```
{-@ ok :: x:Int -> {v:Int | v == x + 3} @-}
ok :: Int -> Int 
ok x = 1 + (2 + x)

{-@ bad :: x:Int -> {v:Int | v == x + 3} @-}
bad :: Int -> Int 
bad = (1 +) . (2 +)
```

The reason the second version fails is that the GHC core looks like:

bad = let t1 = 1 
               t2 = 2
               f   = \a -> a + t1 
               g  = \b -> b + t2
          in
               f . g 

(1) LH needs to infer how the type variables for compose `.` are being 
     instantiated, i.e. that in this case, `.` is being instantiated to:

```
    (a:Int -> {v:Int | v = a + t1}) -> (b:Int -> {v:Int | v = b + t2}) -> c:Int -> {v:Int | v = c + 3}
```     

LH can only do this *if* it has the predicate fragments/qualifiers `v = x + y` 

```
{-@ qualif Plus1(v:Int, x:Int, y:Int): (v == x + y) @-}
```

it all works out. The above are exactly the "output type" for `(+)` but this is 
not used by default due to performance reasons (@gridaphobe -- we need 
to seriously revisit this...) So you can instead run with `--scrape-imports` 
and it has the effect of the above.


The reason it doesn't need these qualifiers in the first case, is that the code, 
in GHC core is:

```
   ok x = let t1 = 2 + x 
                  t2 = 1 + t1 
             in t2
```

And now the inference procedure directly deduces that 

```
t1 :: {v:Int | v = 2 + x }
t2 :: {v:Int | v = 1 + t1}
```

which together *imply* that

```
{v:Int | v == t2} <: {v:Int | v = x + 3}
```

(without having to figure out how the polymorphic types for `.` are being instantiated here.)

For the same reason, if you either add the qualifiers from the refinements in the BS signatures, i.e.

```
{-@ qualif Bs1(obs:ByteString, n:Int): (bslen obs >= n) @-}
{-@ qualif Bs2(obs:ByteString, n:Int): (bslen obs == n) @-}
{-@ qualif Bs3(obs:ByteString, ibs:ByteString, n:Int): (bslen obs == bslen ibs - n) @-}
```

(or turn on `--scrape-imports`) the other version also works fine.

In practice, we've found it convenient to gather commonly used qualifiers 
like the above, and toss them into a "common" library that can be "imported" 
so that they are reused in different modules.

Does that make sense? 

Also: we're currently working on a new version of the inference that 
eliminates the need for these qualifiers (branch `--graphs` and run 
with `--eliminate` but thats rather too "bleeding edge" ATM.)

Does this make sense? Sorry about the wart  -- I was stumped by the
same thing recently elsewhere hence is fresh in my mind...

**Comment by Gabriel439** [comment]
That makes sense!  No apologies necessary :)

**Comment by ranjitjhala** [comment]
Both versions (expanded and non-expanded) now work in `develop` (as the "bleeding edge" `--eliminate` has been promoted to the default.) Thanks for the nice example!
